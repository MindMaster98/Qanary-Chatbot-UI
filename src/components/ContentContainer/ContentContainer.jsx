import React from "react";
import "./ContentContainer.css";

/**
 * 
 * @property {string} id 
 * @property {string} className
 * @property {object} dangerouslySetInnerHTML
 */
export default function ContentContainer(props) {
  return (
    <div
      id={props.id ?? ""}
      className={"contentContainer hidden" + (props.className ?? "")}
      dangerouslySetInnerHTML={
        // This isn't ideal. It would be better to just parse the text to add elements,
        // however since the source is entirely controlled by us, this is fine.
        // Try to avoid this at all costs in production code due to possible XSS
        // vulnerabilities in your application
        !!props.dangerouslySetInnerHTML
          ? {
              __html: props.dangerouslySetInnerHTML,
            }
          : undefined
      }
      tabIndex="0"
      onBlur={(blurEvent) => {
        const self = document.getElementById(props.id);
        // if this element is no longer focused and the new target isn't
        // either part of the element or the header, then hide self
        if (
          !document
            .getElementById("header")
            ?.contains?.(blurEvent?.relatedTarget) &&
          !self.contains(blurEvent?.relatedTarget)
        ) {
          self.classList.add("hidden");
        }
      }}
    >
      {props.children ?? undefined}
    </div>
  );
}
