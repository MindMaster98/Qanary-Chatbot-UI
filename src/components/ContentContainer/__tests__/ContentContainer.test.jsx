import renderer from "react-test-renderer";
import { act } from "@testing-library/react";
import ReactDOM from "react-dom";
import ContentContainer from "../ContentContainer";

// Snapshot tests

it("renders correctly with no props", () => {
  const tree = renderer.create(<ContentContainer />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly with undefined props", () => {
  const tree = renderer.create(<ContentContainer id={undefined} className={undefined} dangerouslySetInnerHTML={undefined} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly with null props", () => {
  const tree = renderer.create(<ContentContainer id={null} className={null} dangerouslySetInnerHTML={null} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly with innerhtml text", () => {
  const tree = renderer.create(<ContentContainer dangerouslySetInnerHTML={"inner html"} />).toJSON();
  expect(tree).toMatchSnapshot();
})

it("renders correctly with innerhtml element", () => {
  const tree = renderer.create(<ContentContainer dangerouslySetInnerHTML={"<button>Element</button>"} />).toJSON();
  expect(tree).toMatchSnapshot();
})

// Blackbox tests
// You cannot dispatch events using test renderer, so I need to actually add the element to the DOM 
// and dispatch an actual event. Calling the function directly would oppose the idea of a blackbox.
// Why not enzyme? Because enzyme doesn't officially support React 17.


let container;

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

it("disappears when focus is lost", () => {
  act(() => {
    ReactDOM.render(<ContentContainer id={"test"} />, container);
  });

  const div = document.querySelector("#test");
  div.classList.remove("hidden");
  // elements need to be focused before being blurrable
  div.focus();
  div.blur();
  
  expect(div.classList.contains("hidden")).toBeTruthy();
});

it("does not disappear when focus is maintained", () => {
  act(() => {
    ReactDOM.render(<ContentContainer id={"test"} />, container);
  });

  const div = document.querySelector("#test");
  div.classList.remove("hidden");
  div.focus();

  expect(div.classList.contains("hidden")).toBeFalsy();
});

it("does not disappear when focus is shifted to another valid element", () => {
  act(() => {
    ReactDOM.render(<ContentContainer id={"test"} dangerouslySetInnerHTML={"<button>Test</button>"} />, container);
  });

  const div = document.querySelector("#test");
  const button = document.querySelector("#test > button");
  div.classList.remove("hidden");
  // elements need to be focused before being blurrable
  div.focus();
  button.focus();

  expect(div.classList.contains("hidden")).toBeFalsy();
});