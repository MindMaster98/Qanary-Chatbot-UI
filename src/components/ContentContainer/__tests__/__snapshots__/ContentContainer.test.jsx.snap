// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with innerhtml element 1`] = `
<div
  className="contentContainer hidden"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<button>Element</button>",
    }
  }
  id=""
  onBlur={[Function]}
  tabIndex="0"
/>
`;

exports[`renders correctly with innerhtml text 1`] = `
<div
  className="contentContainer hidden"
  dangerouslySetInnerHTML={
    Object {
      "__html": "inner html",
    }
  }
  id=""
  onBlur={[Function]}
  tabIndex="0"
/>
`;

exports[`renders correctly with no props 1`] = `
<div
  className="contentContainer hidden"
  id=""
  onBlur={[Function]}
  tabIndex="0"
/>
`;

exports[`renders correctly with null props 1`] = `
<div
  className="contentContainer hidden"
  id=""
  onBlur={[Function]}
  tabIndex="0"
/>
`;

exports[`renders correctly with undefined props 1`] = `
<div
  className="contentContainer hidden"
  id=""
  onBlur={[Function]}
  tabIndex="0"
/>
`;
